### PR TITLE
Autodiscover software

### PR DESCRIPTION
**Pull request by guidow**

### CODE REVIEW

**Comment by guidow** [comment]
The failed test has nothing to do with any changes in this PR, master is broken as well. This failure comes from an API change in psutils. See http://grodola.blogspot.com/2015/06/psutil-30.html, in particular the section "Removal of deprecated APIs".

**Comment by guidow** [comment]
Rebased to master, which now has a fix to the problem mentioned above now.

**Comment by opalmer** [comment]
Can you please add some tests for the new code?  The API endpoint can probably be tested using some of the same tests we use for the other APIs.  For more specific tests you can setup a fake HTTP server to talk back to much like the tests in `TestAgentPostToMaster` or `TestPostShutdownToMaster ` do.

File: pyfarm/agent/http/api/software.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Check Software Endpoint
+===============
```
Comment (opalmer) [inline pyfarm/agent/http/api/software.py]: `=` should be repeated as many times as there are characters in the title.

File: pyfarm/agent/http/api/software.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+try:
+    from httplib import OK, ACCEPTED, INTERNAL_SERVER_ERROR, NO_CONTENT
+except ImportError:  # pragma: no cover
+    from http.client import OK, ACCEPTED, INTERNAL_SERVER_ERROR, NO_CONTENT
+
+from twisted.web.server import NOT_DONE_YET
+from twisted.internet import reactor
+from twisted.internet.defer import Deferred, inlineCallbacks
+
+import treq
```
Comment (opalmer) [inline pyfarm/agent/http/api/software.py]: nit: treq should be imported before twisted

File: pyfarm/agent/http/api/software.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+                    else:
+                        logger.error(
+                            "Failed to remove software %s, version %s from "
+                            "this agent: "
+                            "Unexpected status from server %s. Data: %s",
+                            software, version, response.code, data)
+                        break
+
+        def on_check_software_return(result):
+            if result == True:
```
Comment (opalmer) [inline pyfarm/agent/http/api/software.py]: Either:

```python
if result is True
```

or

```python
if result:
```

File: pyfarm/agent/sysinfo/software.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    from httplib import OK, INTERNAL_SERVER_ERROR
+except ImportError:  # pragma: no cover
+    from http.client import OK, INTERNAL_SERVER_ERROR
+
+import imp, sys
+
+from twisted.internet import reactor, threads
+from twisted.internet.defer import inlineCallbacks, returnValue, Deferred
+
+import treq
```
Comment (opalmer) [inline pyfarm/agent/sysinfo/software.py]: nit: treq should be imported before twisted

File: pyfarm/agent/http/api/software.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+                    logger.error(
+                        "Failed to remove software %s, version %s from this "
+                        "agent on master: %r. Will retry in %s seconds.",
+                        software, version, error, delay)
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                else:
+                    data = yield treq.content(response)
```
Comment (opalmer) [inline pyfarm/agent/http/api/software.py]: `treq.json_content`?

File: pyfarm/agent/http/api/software.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+                        "Failed to post availability of software %s, "
+                        "version %s to master: %r. Will retry in %s "
+                        "seconds.",
+                        software, version, error, delay)
+                    deferred = Deferred()
+                    reactor.callLater(delay, deferred.callback, None)
+                    yield deferred
+
+                else:
+                    data = yield treq.json_content(response)
```
Comment (opalmer) [inline pyfarm/agent/http/api/software.py]: `treq.json_content`?

File: pyfarm/agent/sysinfo/software.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+                logger.error(
+                    "Failed to get discovery code for software %s, version %s: "
+                    "Unexpected status from server %s", software, version,
+                    response.code)
+                query_done = True
+                raise Exception("Unknown return code from master: %s" %
+                                response.code)
+
+@inlineCallbacks
+def check_software_availability(software, version):
```
Comment (opalmer) [inline pyfarm/agent/sysinfo/software.py]: Could we get a doc string here?

File: pyfarm/agent/sysinfo/software.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+                logger.error(
+                    "Failed to get data for software %s, version %s: "
+                    "Unexpected status from server %s", software, version,
+                    response.code)
+                query_done = True
+                raise Exception("Unknown return code from master: %s" %
+                                response.code)
+
+@inlineCallbacks
+def get_discovery_code(software, version):
```
Comment (opalmer) [inline pyfarm/agent/sysinfo/software.py]: Could we get a doc string here?

File: pyfarm/agent/sysinfo/software.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+from pyfarm.agent.http.core.client import get_direct, http_retry_delay
+
+logger = getLogger("agent.sysinfo.software")
+
+class VersionNotFound(Exception):
+    pass
+
+
+@inlineCallbacks
+def get_software_version_data(software, version):
```
Comment (opalmer) [inline pyfarm/agent/sysinfo/software.py]: Could we get a doc string here?

File: pyfarm/agent/sysinfo/software.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+import treq
+
+from pyfarm.agent.logger import getLogger
+from pyfarm.agent.config import config
+from pyfarm.agent.http.core.client import get_direct, http_retry_delay
+
+logger = getLogger("agent.sysinfo.software")
+
+class VersionNotFound(Exception):
```
Comment (opalmer) [inline pyfarm/agent/sysinfo/software.py]: Could we get a doc string here?

File: pyfarm/agent/http/api/software.py (hunk 5 above)
Comment (guidow) [inline pyfarm/agent/http/api/software.py]: Not really necessary at this point. We're not actually doing anything with the data, so there's no need to parse the json content. We're really just looking at the HTTP status code.

File: pyfarm/agent/http/api/software.py (hunk 6 above)
Comment (guidow) [inline pyfarm/agent/http/api/software.py]: Not really necessary at this point. We're not actually doing anything with the data, so there's no need to parse the json content. We're really just looking at the HTTP status code.

**Comment by guidow** [comment]
Okay, tests are added now.

File: pyfarm/agent/http/api/software.py (hunk 5 above)
Comment (opalmer) [inline pyfarm/agent/http/api/software.py]: Alright makes sense.  That should also avoid errors decoding to json too if there happen to be any.

**Comment by opalmer** [comment]
Nice, thanks for adding the tests.  I've rereviewed this and I don't have any additional comments to add so I think this can merge if you think it's ready.

**Pull request merged**